File: CanvasCore/CanvasCore/React Native Modules/NativeLoginManager.m
//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

#import "NativeLoginManager.h"
#import <React/RCTLog.h>
#import <React/RCTBridge.h>
#import <CanvasCore/CanvasCore-Swift.h>

// Object used to send events to React Native about login
@interface NativeLogin : RCTEventEmitter

@property (nonatomic) BOOL isObserving;
@property (nonatomic) NSMutableDictionary *pendingEvents;

@end

static NativeLogin *_sharedInstance;

@implementation NativeLogin

+ (void)setSharedInstance:(NativeLogin *)login {
    _sharedInstance = login;
}

+ (NativeLogin *)sharedInstance {
    return _sharedInstance;
}

- (instancetype)init {
    self = [super init];
    self.isObserving = NO;
    self.pendingEvents = [NSMutableDictionary new];
    [NativeLogin setSharedInstance:self];
    return self;
}

RCT_EXPORT_MODULE();

- (NSDictionary *)constantsToExport
{
    return @{ @"isTesting": @([NSProcessInfo processInfo].environment[@"IS_UI_TEST"] != nil) };
}

RCT_EXPORT_METHOD(changeUser)
{
    [[NativeLoginManager shared].delegate changeUser];
}

RCT_EXPORT_METHOD(stopActing)
{
    [[NativeLoginManager shared].delegate stopActing];
}

RCT_EXPORT_METHOD(logout)
{
    [[NativeLoginManager shared].delegate logout];
}

RCT_EXPORT_METHOD(actAsFakeStudentWithID:(NSString *)fakeStudentID)
{
    [[NativeLoginManager shared].delegate actAsFakeStudentWithID:fakeStudentID];
}

RCT_EXPORT_METHOD(startObserving)
{
    self.isObserving = YES;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(.25 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self.pendingEvents enumerateKeysAndObjectsUsingBlock:^(NSString *key, id obj, BOOL * _Nonnull stop) {
            [self sendEventWithName:key body:obj];
        }];
        [self.pendingEvents removeAllObjects];
    });
}

RCT_EXPORT_METHOD(stopObserving)
{
    self.isObserving = NO;
}

- (dispatch_queue_t)methodQueue { return dispatch_get_main_queue(); }
+ (BOOL)requiresMainQueueSetup { return YES; }
- (NSArray<NSString *> *)supportedEvents { return @[@"Login"]; }

- (void)sendEventWithName:(NSString *)name body:(id)body {
    if (self.isObserving) {
        [super sendEventWithName:name body:body];
    }
    else {
        self.pendingEvents[name] = body;
    }
}

@end

@implementation NativeLoginManager

+ (instancetype)shared {
    static NativeLoginManager *manager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        manager = [[NativeLoginManager alloc] init];
    });
    return manager;
}

- (id)init {
    self = [super init];
    return self;
}

- (void)login:(NSDictionary *)body {
    [[NativeLogin sharedInstance] sendEventWithName:@"Login" body:body];
}

- (void)logout {
    [[HelmManager shared] cleanupWithCallback:^{
        [[NativeLogin sharedInstance] sendEventWithName:@"Login" body:@{}];
    }];
}

@end

File: CanvasCore/CanvasCore/ReactNativeModules/NativeLoginManager.swift
//
// This file is part of Canvas.
// Copyright (C) 2019-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import Foundation
import Core

extension NativeLoginManager {
    public static func login(as entry: LoginSession, brand: Core.Brand = .shared) {
        var body: [String: Any] = [
            "appId": Bundle.main.isTeacherApp ? "teacher" : "student",
            "authToken": entry.accessToken ?? "",
            "refreshToken": entry.refreshToken ?? "",
            "clientID": entry.clientID ?? "",
            "clientSecret": entry.clientSecret ?? "",
            "baseURL": entry.baseURL.absoluteString,
            "branding": [
                "buttonPrimaryBackground": brand.buttonPrimaryBackground.hexString,
                "buttonPrimaryText": brand.buttonPrimaryText.hexString,
                "buttonSecondaryBackground": brand.buttonSecondaryBackground.hexString,
                "buttonSecondaryText": brand.buttonSecondaryText.hexString,
                "fontColorDark": brand.fontColorDark.hexString,
                "headerImageBackground": brand.headerImageBackground.hexString,
                "linkColor": brand.linkColor.hexString,
                "navBackground": brand.navBackground.hexString,
                "navBadgeBackground": brand.navBadgeBackground.hexString,
                "navBadgeText": brand.navBadgeText.hexString,
                "navIconFill": brand.navIconFill.hexString,
                "navIconFillActive": brand.navIconFillActive.hexString,
                "navTextColor": brand.navTextColor.hexString,
                "navTextColorActive": brand.navTextColorActive.hexString,
                "primary": brand.primary.hexString,
            ],
            "countryCode": Locale.current.regionCode ?? "",
            "locale": LocalizationManager.currentLocale ?? "en",
            "user": [
                "avatar_url": entry.userAvatarURL?.absoluteString,
                "id": entry.userID,
                "name": entry.userName,
                "primary_email": entry.userEmail,
            ],
            "isFakeStudent": entry.isFakeStudent,
            "isK5Enabled": AppEnvironment.shared.k5.isK5Enabled,
        ]
        if let actAsUserID = entry.actAsUserID {
            body["actAsUserID"] = actAsUserID
        }
        shared().login(body)
    }
}
